lib/application.js

```diff
@@ -567,3 +571,65 @@ app.listen = function(cb) {
 
   return server;
 };
+
+/* START CONTROLLER CODE*/
+
+// this is similar to app.model
+// at the moment I haven't really used 'options'
+app.controller = function(ctor, options) {
+  var self = this;
+
+  // the SharedController is a wrapper, see below
+  var controller = new SharedController(ctor);
+  this.remotes().addClass(controller.sharedClass);
+  app.controllers().push(controller);
+
+  return controller;
+
+  // SharedController class contains strong-remoting objects to set up the metadata
+  // and register the endpoint, etc
+  function SharedController(ctor, options) {
+    options = options || {};
+    this.name = ctor.name;
+    this.ctor = ctor;
+    // TODO - right now there's a hack on line 289 to skip this
+    // util.inherits(ctor, EventEmitter);
+
+    this.sharedClass = new RemoteObjects.SharedClass(this.name, ctor, options);
+    // the resolveScope method instantiates the controller upon request
+    // (please correct my terminology if it's wrong)
+    this.sharedClass.resolveScope = function(remotingContext) {
+      // self is the app - should it be something more obvious like appInstance?
```

Comment:
I think this shows that the current design, where a new function `SharedController` is defined for each `app.controller` call, is suboptimal.

I would prefer if `SharedController` was a standalone class, defining methods on a prototype, and receiving the `app` or `appInstance` object via constructor.

```js
// goes to lib/shared-controller.js
function SharedController(ctor, app, options) {
  // ...
  this.sharedClass.resolveScope = function(remotingContext) {
     return new ctor(app, remotingContext);
  };
}

SharedController.prototype.remoteMethod = function(name, options) {
  // ...
};

// etc.
```